api/src/main/java/org/pmiops/workbench/profile/ProfileService.java

```diff
@@ -286,50 +287,132 @@ private void updateInstitutionalAffiliations(Profile updatedProfile, DbUser user
+    if (username == null || username.length() < 3 || username.length() > 64) {
+      throw new BadRequestException(
+          "Username should be at least 3 characters and not more than 64 characters");
     }
-    Optional.ofNullable(updatedProfile.getAddress())
+  }
+
+  private void validateContactEmail(Profile profile) throws BadRequestException {
+    if (StringUtils.isEmpty(profile.getContactEmail())) {
+      throw new BadRequestException("Contact email cannot be empty");
+    }
+  }
+
+  private void validateGivenName(Profile profile) throws BadRequestException {
+    validateStringLength(profile.getGivenName(), "Given Name", 80, 1);
+  }
+
+  private void validateFamilyName(Profile profile) throws BadRequestException {
+    validateStringLength(profile.getFamilyName(), "Family Name", 80, 1);
+  }
+
+  private void validateAddress(Profile profile) throws BadRequestException {
+    Optional.ofNullable(profile.getAddress())
         .orElseThrow(() -> new BadRequestException("Address must not be empty"));
 
-    Address updatedProfileAddress = updatedProfile.getAddress();
+    Address updatedProfileAddress = profile.getAddress();
     if (StringUtils.isEmpty(updatedProfileAddress.getStreetAddress1())
         || StringUtils.isEmpty(updatedProfileAddress.getCity())
         || StringUtils.isEmpty(updatedProfileAddress.getState())
         || StringUtils.isEmpty(updatedProfileAddress.getCountry())
         || StringUtils.isEmpty(updatedProfileAddress.getZipCode())) {
       throw new BadRequestException(
-          "Address cannot have empty street Address 1/city/state/country or Zip Code");
+          "Address cannot have empty street address, city, state, country or zip code");
     }
-    if (updatedProfile.getContactEmail() != null
-        && !updatedProfile.getContactEmail().equals(prevProfile.getContactEmail())) {
-      // See RW-1488.
-      throw new BadRequestException("Changing email is not currently supported");
-    }
-    if (updatedProfile.getUsername() != null
-        && !updatedProfile.getUsername().equals(prevProfile.getUsername())) {
-      // See RW-1488.
-      throw new BadRequestException("Changing username is not supported");
+  }
+
+  private void validateAreaOfResearch(Profile profile) throws BadRequestException {
+    if (StringUtils.isEmpty(profile.getAreaOfResearch())) {
+      throw new BadRequestException("Research background cannot be empty");
     }
   }
 
-  private void validateStringLength(String field, String fieldName, int max, int min) {
-    if (field == null) {
-      throw new BadRequestException(String.format("%s cannot be left blank!", fieldName));
+  /**
+   * Returns a list of PropertyChange entries from a diff that match either a field or any of its
+   * subfields. For example, a pathPrefix of 'address' will match a field change for both 'address'
+   * and for 'address.zipCode'.
+   */
+  private List<Change> getChangesWithPrefix(final Diff diff, final String pathPrefix) {
```

Comment:
Is there not a way to associate a schema with a Diff? It seems somewhat risky (in terms of unintended changes in behavior) to use reflection, though I can see the other argument.

Reply:
I thought of a couple of things to be careful about would be any dynamic properties that return different results every time or properties derived from other properties, which would be double-evaluated. Neiither are an issue for a simple codegen'd model class I suppose.